sbt-launcher/src/main/scala/coursier/sbtlauncher/implem/Launcher.scala

```diff
@@ -114,8 +110,14 @@ class Launcher(
       }
     )
 
-  lazy val topLoader: ClassLoader =
-    classOf[xsbti.Launcher].getClassLoader
+  private class TestInterfaceLoader(urls: Array[URL], parent: ClassLoader)
+      extends URLClassLoader(urls, parent)
+  lazy val topLoader: ClassLoader = {
+    val top = classOf[xsbti.Launcher].getClassLoader
+    sbtTestInterfaceFilesOpt.fold(top) { sbtTestInterfaceFiles =>
+      new TestInterfaceLoader(sbtTestInterfaceFiles.map(_.toURI.toURL).toArray, top)
```

Comment:
Couldn't there be a more reliable way to detect that in sbt? Like just checking if things from test-interface and sbt-specific stuff come from two different classloader?

I'm planning on maybe trying to run this launcher from graalvm; it would start sbt by running a `java -cp` command, maybe using things from coursier bootstrap to get the classloader hierarchy when the `java` command starts. Having it give its classloaders' classes specific names seems pretty ad hoc, and would require quite a bit of circonvolutions (adjusting a class name at runtime…).

Reply:
Sure, there could be a better way of detecting it. I don't have a ton of motivation to change it though because the vast majority of sbt users are using the default sbt launcher.

For the coursier launcher, I think the sensible choice is to either make short circuit default for sbt 1.3.0 or go with this patch. It doesn't hurt anything and I don't expect sbt to change the expected class loader layout. Right now, the coursier launcher does not work optimally with sbt 1.3.0 because sbt will discard the input classloader and create a new classloader to launch itself.

As long as the launcher guarantees test interface loader appears above the scala and sbt classloaders, it is fine to bypass xMain.run and call xMain$.run instead. Things go wrong in sbt >= 1.3.0 if that assumption is violated so sbt aggressively discards any classloader that it can't guarantee has that layout.